File: services/mirror-worker/src/replay/event-justified.ts
import axios from 'axios';
import { PoolClient } from 'pg';
import { config } from '../config';
import { log } from '../log';
import { ProdUserLog } from '../db/prod';
import { insertMapping } from '../db/staging';
import { getStagingApiToken } from '../runtime';
import { translateEquipmentEventId, translateEquipmentId } from '../translate';

// Prod payload shape captured from session 55 inspection:
//   {
//     idle: 'no', cdMachine, cdCategory, descCategory,
//     cdSubcategory, descSubcategory, changeOver, plannedDowntime,
//     idEquipment: <prod>,
//     idEquipmentEvent: <prod>,
//     txtDowntimeNotes
//   }
//
// Translation:
//   idEquipment      -> via packml_topic
//   idEquipmentEvent -> via business key (id_equipment + ts_event + status)
//                       cached in mirror_id_map after first hit
//
// Output: POST /api/downtimes/justify with idEnterprise added
// (operator UI passes it as a query param on prod; we send it in the body
// for consistency with the worker pipeline).
interface JustifyPayload {
  idle: string;
  cdMachine?: string;
  cdCategory?: string;
  descCategory?: string;
  cdSubcategory?: string;
  descSubcategory?: string;
  changeOver?: boolean;
  plannedDowntime?: boolean;
  idEquipment: number;
  idEquipmentEvent: number;
  txtDowntimeNotes?: string;
}

export async function replayEventJustified(
  client: PoolClient,
  row: ProdUserLog,
): Promise<void> {
  const payload = row.payload as JustifyPayload;
  if (!payload || typeof payload !== 'object') {
    throw new Error('event-justified payload missing or not an object');
  }
  if (typeof payload.idEquipment !== 'number') {
    throw new Error('event-justified payload missing idEquipment');
  }
  if (typeof payload.idEquipmentEvent !== 'number') {
    throw new Error('event-justified payload missing idEquipmentEvent');
  }

  const stagingEquipmentId = await translateEquipmentId(payload.idEquipment);
  const stagingEventId = await translateEquipmentEventId(
    BigInt(payload.idEquipmentEvent),
  );
  if (stagingEventId === undefined) {
    throw new Error(
      `equipment_event ${payload.idEquipmentEvent} unmapped (no business-key match within 5s window)`,
    );
  }

  // Cache the equipment_event mapping after the first business-key resolution
  // so future event-edited / event-splitted for the same event don't repeat
  // the lookup.
  await insertMapping(client, {
    entityType: 'equipment_event',
    source: config.sourceName,
    prodId: BigInt(payload.idEquipmentEvent),
    stagingId: stagingEventId,
    sourceLogId: BigInt(row.id_user_logs),
  });

  const body = {
    ...payload,
    idEquipment: stagingEquipmentId,
    idEquipmentEvent: Number(stagingEventId),
    idEnterprise: config.stagingEnterpriseId,
  };

  const url = `${config.stagingApi.baseUrl}/api/downtimes/justify`;
  const res = await axios.post(url, body, {
    params: {
      token: getStagingApiToken(),
      idEnterprise: config.stagingEnterpriseId,
    },
    headers: {
      'x-user': row.nm_user ?? 'mirror-worker',
      'X-Mirror-Source': config.sourceName,
    },
    validateStatus: () => true,
    timeout: 15_000,
  });

  if (res.status >= 400) {
    throw new Error(
      `staging /api/downtimes/justify returned ${res.status}: ${JSON.stringify(res.data)}`,
    );
  }

  log.info(
    {
      sourceLogId: row.id_user_logs,
      prodEventId: payload.idEquipmentEvent,
      stagingEventId: stagingEventId.toString(),
      status: res.status,
    },
    'replayed event-justified',
  );
}

File: services/mirror-worker-go/internal/replay/event_justified.go
package replay

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/jackc/pgx/v5"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/config"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/db"
	"github.com/packiot/packiot-stack-alpha/services/mirror-worker-go/internal/translate"
)

// JustifyPayload mirrors prod's shape. Forwarded almost verbatim to staging
// — we replace idEquipment / idEquipmentEvent with their staging IDs and
// add idEnterprise. Unknown extra fields are tolerated by class-validator.
type JustifyPayload struct {
	Idle             string  `json:"idle,omitempty"`
	CdMachine        string  `json:"cdMachine,omitempty"`
	CdCategory       string  `json:"cdCategory,omitempty"`
	DescCategory     string  `json:"descCategory,omitempty"`
	CdSubcategory    string  `json:"cdSubcategory,omitempty"`
	DescSubcategory  string  `json:"descSubcategory,omitempty"`
	ChangeOver       *bool   `json:"changeOver,omitempty"`
	PlannedDowntime  *bool   `json:"plannedDowntime,omitempty"`
	IDEquipment      int     `json:"idEquipment"`
	IDEquipmentEvent int64   `json:"idEquipmentEvent"`
	TxtDowntimeNotes string  `json:"txtDowntimeNotes,omitempty"`
}

func EventJustified(
	cfg *config.Config,
	t *translate.Translator,
	apiToken func() string,
	httpc *http.Client,
	logger *slog.Logger,
) Handler {
	return func(ctx context.Context, tx pgx.Tx, row db.ProdUserLog) error {
		var p JustifyPayload
		if err := json.Unmarshal(row.Payload, &p); err != nil {
			return fmt.Errorf("parse event-justified payload: %w", err)
		}
		if p.IDEquipment == 0 || p.IDEquipmentEvent == 0 {
			return fmt.Errorf("event-justified payload missing fields: %+v", p)
		}

		stagingEqID, err := t.Equipment(ctx, p.IDEquipment)
		if err != nil {
			return fmt.Errorf("translate equipment: %w", err)
		}
		stagingEventID, err := t.EquipmentEvent(ctx, p.IDEquipmentEvent)
		if err != nil {
			if errors.Is(err, translate.ErrUnmapped) {
				return fmt.Errorf("equipment_event %d unmapped (no staging row with >=%ds interval overlap; see WARN log): %w",
					p.IDEquipmentEvent, cfg.EventMinOverlapSec, err)
			}
			return fmt.Errorf("translate equipment_event: %w", err)
		}

		// Cache the equipment_event mapping after the first business-key
		// resolution so future event-edited / event-splitted on the same
		// prod event skip the matcher SELECT.
		if err := db.InsertMapping(ctx, tx, db.MapInsert{
			EntityType:  "equipment_event",
			Source:      cfg.SourceName,
			ProdID:      p.IDEquipmentEvent,
			StagingID:   stagingEventID,
			SourceLogID: row.IDUserLogs,
		}); err != nil {
			return fmt.Errorf("insert mapping: %w", err)
		}

		// Pass payload through verbatim, override the translated fields.
		// json.RawMessage round-trip preserves any extra fields prod sent.
		var raw map[string]any
		if err := json.Unmarshal(row.Payload, &raw); err != nil {
			return fmt.Errorf("re-parse payload: %w", err)
		}
		raw["idEquipment"] = stagingEqID
		raw["idEquipmentEvent"] = stagingEventID
		raw["idEnterprise"] = cfg.StagingEnterpriseID

		status, _, err := PostStaging(ctx, cfg, httpc, apiToken(), row,
			"/api/downtimes/justify", raw)
		if err != nil {
			return err
		}
		logger.Info("replayed event-justified",
			slog.Int64("sourceLogID", row.IDUserLogs),
			slog.Int64("prodEventID", p.IDEquipmentEvent),
			slog.Int64("stagingEventID", stagingEventID),
			slog.Int("status", status))
		return nil
	}
}
